code/modules/mafia/roles/roles.dm

```diff
@@ -46,11 +51,43 @@
 		role_unique_actions -= abilities
 
 /datum/mafia_role/Destroy(force, ...)
+	send_back_to_body()
 	QDEL_NULL(mafia_alert)
 	QDEL_NULL(body)
-	QDEL_NULL(role_unique_actions)
+	QDEL_LIST(role_unique_actions)
 	return ..()
 
+/**
+ * Puts the player in their body and keeps track of their previous one to put them back in later.
+ * Adds the playing_mafia trait so people examining them will know why they're currently lacking a soul.
+ */
+/datum/mafia_role/proc/put_player_in_body(client/player)
+	if(player.mob.mind && player.mob.mind.current)
+		old_body = player.mob.mind.current
+		old_mind = player.mob.mind
+		ADD_TRAIT(old_body, TRAIT_PLAYING_MAFIA, MAFIA_TRAIT)
+	body.key = player.key
+
+/**
+ * Sends the player back into their body
+ * Ghostize the body to get the observer. If you get nothing, they died in-game and is a ghost,
+ * so get the body's ghost instead.
+ * Once we have the ghost, we set their mind to their old mind, then set the current
+ * If the player is alive again, we'll shove them back in.
+ */
+/datum/mafia_role/proc/send_back_to_body()
+	if(old_mind && !QDELETED(old_mind) && old_body && !QDELETED(old_body))
```

Comment:
QDELETED does a nullcheck, so this is just pretty redundant, should be:

```suggestion
	if(!QDELETED(old_mind) && !QDELETED(old_body))
```

also this could probably be an early return instead to save that indentation